workflow/session.go

```diff
@@ -0,0 +1,112 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package workflow
+
+import (
+	"go.uber.org/cadence/internal"
+)
+
+// SessionInfo contains information of a created session. For now, the only exported
+// field is SessionID which has string type. SessionID is a uuid generated when
+// CreateSession() or RecreateSession() is called and can be used to uniquely identify
+// a session.
+type SessionInfo = internal.SessionInfo
+
+// Note: Worker should be configured to process session. To do this, set the following
+// fields in WorkerOptions:
+//     EnableSessionWorker: true
+//     SessionResourceID: The identifier of the resource consumed by sessions.
+//         It's the user's responsibility to ensure there's only one worker using this resourceID.
+//     MaxConCurrentSessionExecutionSize: the maximum number of concurrently sessions the resource
+//         support. By default, 1000 is used.
+
+// CreateSession creates a session and return a new context which contains information
+// of the created session. The session will be created on the tasklist user specified in
+// ActivityOptions. If none is specified, the default one will be used.
+//
+// CreationSession will fail in the following situations:
+//     1. The context passed in already contains a session which is still open
+//        (not closed and failed).
+//     2. The number of sessions currently running on the worker has reached
+//        MaxConCurrentSessionExecutionSize, which is specified when starting the worker.
+//
+// If an activity is executed using the returned context, it's regarded as part of the
+// session and all activities within a session will be executed by the same worker.
+// User still needs to handle the error returned when executing an activity. Session will
+// not be marked as failed if an activity within it returns an error. Only when the worker
+// executing the session is down, that session will be marked as failed. Executing an activity
+// within a failed session will return an [TODO] error immediately without even scheduling the activity.
+//
+// If user wants to end a session since some activity returns an error, use the CompleteSession API below,
+// and new session can be created if necessary to retry the whole session.
+//
+// Example:
+//    sessionCtx, err := CreateSession(ctx)
+//    if err != nil {
+//		    // Creation failed. Wrong ctx or too many outstanding sessions.
+//    }
+//    err = ExecuteActivity(sessionCtx, someActivityFunc, activityInput).Get(sessionCtx, nil)
+//    if err != nil {
+//        // Session has failed or activity itself failed.
+//    }
+//    ... // execute more activities using sessionCtx
+//    err = CompleteSession(sessionCtx)
+//    if err != nil {
+//        // Wrong ctx is used or failed to release session resource.
+//    }
+func CreateSession(ctx Context) (Context, error) {
```

Comment:
Let's add a SessionOptions struct as a parameter. For now let's add creationTimeout (which maps to stcheduleToStart sessionCreationActivity timeout) to it.

Reply:
Done. SessionOptions currently has two fields: creationTimeout and executionTimeout.